### PR TITLE
i32: fused Q31 gain GainQ31 with pre-shift and rounding requant (Fixes #249)

### LINKED ISSUE #249: i32: fused Q31 gain with pre-shift and rounding post-shift (denormaliseBands requant)
## Summary

Extend the `i32` fixed-point surface with a fused vector-by-scalar Q31 gain that also applies a per-call left pre-shift and a rounding right post-shift. `ScaleQ31` from #181 covers the bare `MULT32_32_Q31(x, g)` core, but the gain-application hot path in a fixed-point audio decoder wraps that core in an input `SHL32` and an output `PSHR32`, and today those two shifts have to run as separate passes over the buffer.

## The op

Per element, with per-call scalar constants `g` (Q31 gain), `preShift`, and `postShift`:

```
dst[i] = PSHR32(MULT32_32_Q31(SHL32(src[i], preShift), g), postShift)
```

The `MULT32_32_Q31` core is exactly what `ScaleQ31` already implements, so a fused kernel stays bit-identical to the `ScaleQ31` + shift composition. `SHL32` is a plain left shift. `PSHR32(_, n)` is a round-half-up arithmetic shift (add `1 << (n-1)` before shifting), not a truncation.

## Motivation

This is the `denormaliseBands` inner loop in the go-opus fixed-point Opus port (tphakala/go-opus), about 3.5 percent of a stereo 20 ms decode at 48 kHz. #181 already named `denormaliseBands` as motivation and shipped `ScaleQ31`, but left the surrounding `SHL32` and `PSHR32` to the caller, so the caller still makes three passes (pre-shift, scale, round-shift) where one fused kernel would do the work in a single sweep. Input pre-shift and rounding output requant are general fixed-point gain staging, not Opus-specific, so this sits naturally next to `ScaleQ31`.

## Notes

- `postShift == 0` must skip the rounding bias entirely (no `1 << -1`); the caller does pass 0.
- The caller also clamps `g` to `2147483647` and forces `shift = 0` for extreme gains, so a plain (non-saturating, wrapping int32) kernel that matches `ScaleQ31` is correct here; there is no relaxed tier and the pure-Go and SIMD paths stay bit-identical.
- Sibling of #181. The Opus-side adoption in `denormaliseBands` is a follow-up on the tphakala/go-opus side once this lands.

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

`GainQ31` adds a fused fixed-point Q31 gain to package `i32`: per element, `dst[i] = PSHR32(MULT32_32_Q31(SHL32(a[i], preShift), g), postShift)`, applied in a single pass. It wraps the existing `ScaleQ31` core (`MULT32_32_Q31`) in an input `SHL32` pre-shift and a rounding `PSHR32` output requant, the exact shape the fixed-point Opus `denormaliseBands` gain-application loop applies around it, so a caller collapses three passes (pre-shift, scale, round-shift) into one sweep.

Every stage wraps in int32 (no saturation), so the AVX2, NEON and pure-Go paths are bit-identical and the result equals the `ScaleQ31` + `SHL32`/`PSHR32` composition exactly. The pre-shift runs on the int32 value before the widen, keeping the product within 2^62 (no int64 overflow), and the rounding bias `(int32(1)<<postShift)>>1` is computed in int32 arithmetic so it matches the reference `PSHR32` bit-for-bit, including the `postShift == 31` sign-extension corner. `preShift` and `postShift` must be in `[0, 31]`.

## Implementation

- Go reference `gainQ31Go`, which masks the shift counts into `[0,31]` once so the hot loop drops the `runtime.panicshift` and range-clamp guards.
- AVX2 `gainQ31AVX2`: `VPSLLD` pre-shift, the `scaleQ31AVX2` `VPMULDQ` Q31 core verbatim, `VPADDD` bias, `VPSRAD` post-shift.
- NEON `gainQ31NEON`: `SSHL` pre-shift, the `scaleQ31NEON` `SMULL`/`SSHR`/`XTN` core verbatim, `ADD` bias, negative-count `SSHL` arithmetic post-shift (single rounding).
- Independent arbitrary-precision `big.Int` oracle plus parity, wrap, per-lane value-matrix, aliasing, unaligned, allocation-free and fuzz tests. Every hand-encoded NEON `WORD` is cross-checked against `golang.org/x/arch/arm64asm`.

## Related Issues

Fixes #249

## Test Plan

- [x] Full module suite green on amd64 (`go test ./...`) plus the assembly-encoding checks
- [x] Full `i32` suite green on native arm64 (Raspberry Pi 5)
- [x] 1.2M+ differential fuzz executions (AVX2/NEON vs Go reference vs independent oracle)
- [x] Allocation-free; `golangci-lint` clean on both build tags
- [x] Benchmarks allocation-free: AVX2 about 5x and NEON about 3x over the scalar path at n=1000

## Summary by CodeRabbit

- **New Features**
  - Added the public `GainQ31` operation for fused fixed-point gain processing on 32-bit integer data.
  - Supports pre-shifting, Q31 multiplication, rounding, post-shifting, in-place processing, and length-safe output handling.
  - Added SIMD acceleration on supported x86-64 and ARM64 systems, with automatic fallback for other environments.

- **Documentation**
  - Documented `GainQ31` and its processing behavior in the API documentation.

- **Tests**
  - Added comprehensive correctness, edge-case, allocation, aliasing, and SIMD coverage.

### CODE REVIEW

File: i32/gain.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// the samples can be processed in place. dst must not otherwise overlap a: the
+// SIMD kernels load a whole block of a before storing the block of dst, so a
+// shifted dst/a overlay could clobber input lanes a later iteration has not read
+// yet. This is the same aliasing rule and caveat as [ScaleQ31].
+func GainQ31(dst, a []int32, g int32, preShift, postShift int) {
+	n := min(len(dst), len(a))
+	if n == 0 {
+		return
+	}
+	gainQ31I32(dst[:n], a[:n], g, preShift, postShift)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply i32/gain.go:53]: Good call. Added the [0,31] validation to the public GainQ31 in 96f6ee1: it panics before the length check if preShift or postShift is out of range, matching the convention that public functions validate their inputs. Boundary tests cover the negative, >31, both-invalid, and empty-slice cases plus the [0,31] endpoints. The check is once per call, so the hot loops are untouched.

File: README.md [hunk 2]
```diff
@@ -631,7 +632,7 @@ i32.ScaleQ15(dst, left, 16384) // truncating Q15 scale, 16384 = 0.5 in Q15
...
...the pure-Go reference by construction (~10x AVX2, ~5x NEON). All zero-allocation. The fixed-point building blocks `ScaleQ31` and `ScaleQ15` are truncating scale-by-scalar multiplies (the integer `MULT32_32_Q31` and `MULT16_32_Q15`: a 64-bit product arithmetically shifted back into int32 with no rounding constant), `GainQ31` fuses that `MULT32_32_Q31` core with an input `SHL32` pre-shift and a rounding `PSHR32` output requant in a single pass (the integer-Opus denormalise-bands gain application, round half up), `Butterfly` is the Haar/FWHT radix-2 combine (`lo, hi = lo+hi, lo-hi`), and `FIRValidQ15` is the int32 valid convolution against int16 Q15 taps, quantized per product (not once at the end); all truncate rather than round and carry no rounding constant. `MaxAbs` is the `celtMaxabs32` peak magnitude (`max(maxVal, -minVal)`) built on the same signed `MinMax` scan, and `NegWhereNeg` is a branchless conditional negate driven by a parallel float32 sign stream. Every one wraps in int32 (no saturation), is bit-exact across amd64 AVX2, arm64 NEON and pure Go with no relaxed tier, and allocation-free: these are the integer-DSP and fixed-point-codec (integer Opus, FWHT) building blocks.
```
[the remark this answers is not in the document]
Comment (tphakala) [reply README.md]: Fixed in 96f6ee1. Reworded the closing clause so only ScaleQ31, ScaleQ15, and FIRValidQ15 are described as truncating with no rounding constant, and called out GainQ31 as applying the PSHR32 rounding bias.

File: i32/gain.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply i32/gain.go:54]: Right about the NEON SSHL divergence: a negative count there becomes an arithmetic right shift, while gainQ31Go masks to 31 and the amd64 scalar tail masks to 5 bits, so out-of-range shifts would disagree across backends. Fixed in 96f6ee1 by validating preShift/postShift to [0,31] at the entry point (panic otherwise), so no out-of-range count reaches any backend.
